Add PathParamExtractor

lespinsideg opened the pull request:
issue number  : #421 

I wrote PathParamExtractor that extracts path param values from HttpRequest and test cases for that.
Please review them.

imasahiro commented on core/src/main/java/com/linecorp/armeria/common/util/PathParamExtractor.java:
```diff
@@ -0,0 +1,93 @@
+ */
+public class PathParamExtractor {
+    @SuppressWarnings("unchecked")
+    private static final Map<String, String> EMPTY_MAP = Collections.EMPTY_MAP;
```
Please use Collections.emptyMap().

imasahiro commented on core/src/test/java/com/linecorp/armeria/common/util/PathParamExtractorTest.java:
```diff
@@ -0,0 +1,71 @@
+        assertThat(values.get("value2")).isEqualTo("world");
+    }
+
+    @Test(expected = IllegalArgumentException.class)
```
Could you use `assertThatThrownBy(() -> new PathParamExtractor(...))` instead?

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/PathParamExtractor.java:
```diff
@@ -0,0 +1,93 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common.util;
```
I wonder this class could be useful for non-HTTP. If not, how about moving it to `com.linecorp.armeria.server.http`? If this is useful on the client side as well, we could place it at `com.linecorp.armeria.common.http`.

lespinsideg commented:
@trustin Okay, I will merge DynamicPath then. 
@imasahiro @trustin Thank you for the review. I will reflect them after merging DynamicPath.

lespinsideg commented:
@trustin I am sorry for late update. I renamed DynamicPath to PathParamExtractor and modified it.
- if there is no matching pattern, extract method returns empty map instead of null
- more restrict VALID_PATTERN

trustin commented on core/src/main/java/com/linecorp/armeria/common/http/PathParamExtractor.java:
```diff
@@ -26,38 +28,41 @@
- *   <li> A regex-compiled form of the path. It is used for matching and binding.
+ *   <li> A regex-compiled form of the path. It is used for matching and extracting.
  *   <li> A skeleton of the path. It is used for duplication detecting.
  *   <li> A set of variables declared in the path.
```
Would you mind:

- Removing the leading space for each list item and
- Closing the `<li>` tag

.. although it's not your change strictly speaking?

trustin commented on core/src/main/java/com/linecorp/armeria/common/http/PathParamExtractor.java:
```diff
@@ -26,38 +28,41 @@
     /**
-     * Create a {@link DynamicPath} instance from given {@code path}.
+     * Create a {@link PathParamExtractor} instance from given {@code path}.
+     * @param path the {@link String} that contains path params.
```
Please insert a new line between the main description and the `@param` tag. i.e.

```java
/**
 * ...
 *
 * @param path ...
 *
 * @throws ...
 */
```

trustin commented on core/src/main/java/com/linecorp/armeria/common/http/PathParamExtractor.java:
```diff
@@ -72,11 +77,11 @@ static DynamicPath of(String path) {
     private final ImmutableSet<String> variables;
 
-    DynamicPath(String path) {
+    private PathParamExtractor(String path) {
```
I wonder if we could let users just use the public constructor given that `of()` just delegates to this constructor. i.e. How about removing `of()` and making this constructor public?

trustin replied:
path -> pathPattern

trustin commented on core/src/main/java/com/linecorp/armeria/common/http/PathParamExtractor.java:
```diff
@@ -101,7 +106,7 @@ static DynamicPath of(String path) {
         }
 
-        pattern = Pattern.compile(patternJoiner.toString());
+        pattern = Pattern.compile(patternJoiner.toString() + "(\\?.*)*");
```
`.toString()` is redundant.

trustin commented on core/src/main/java/com/linecorp/armeria/common/http/PathParamExtractor.java:
```diff
@@ -26,38 +28,41 @@
- * @see Path
  */
-final class DynamicPath {
+public final class PathParamExtractor {
```
Could you implement equals/hashCode/toString?

trustin commented on core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionEntry.java:
```diff
@@ -63,11 +65,11 @@
 
     /**
-     * Returns bound values and mapped function when given HTTP Method and path matches.
+     * Returns bound values and mapped function when given HTTP Method and pathParamExtractor matches.
```
pathParamExtractor -> `{@link PathParamExtractor}`

trustin commented on core/src/test/java/com/linecorp/armeria/common/http/PathParamExtractorTest.java:
```diff
@@ -0,0 +1,87 @@
+        Map<String, String> values = ppe.extract("/service/hello");
+
+        assertThat(values).hasSize(1);
+        assertThat(values.get("value")).isEqualTo("hello");
```
How about:

```java
assertThat(values).containsOnlyKeys("value").containsValues("hello");
```

trustin commented on core/src/test/java/com/linecorp/armeria/common/http/PathParamExtractorTest.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Test
+    public void testMultipleMatch() throws Exception {
```
Match -> Matches

trustin commented on core/src/test/java/com/linecorp/armeria/common/http/PathParamExtractorTest.java:
```diff
@@ -0,0 +1,87 @@
+
+        Map<String, String> values = ppe.extract("service/hello/test/world/something?q=1");
+
+        assertThat(values).hasSize(2);
```
How about: `assertThat(values).containsOnlyKeys("value", "value2");`

trustin commented on core/src/test/java/com/linecorp/armeria/common/http/PathParamExtractorTest.java:
```diff
@@ -0,0 +1,87 @@
+
+        assertThatThrownBy(() -> PathParamExtractor.of("/service/:value/test/value2}"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("/service/:value/test/value2}: Invalid PathParamExtractor");
```
Could you change the exception message to:

```java
throw new IllegalArgumentException("pathPattern: " + pathPattern);
```

.. in `PathParamExtractor`?

trustin commented on core/src/test/java/com/linecorp/armeria/common/http/PathParamExtractorTest.java:
```diff
@@ -0,0 +1,87 @@
+    public void testEmptyPattern() throws Exception {
+        assertThatThrownBy(() -> PathParamExtractor.of("/service/value/test/value2"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("No placeholder exists in given path: /service/value/test/value2");
```
I'm curious if a path pattern without variable placeholders should be rejected. Perhaps OK to return an empty map on `extract()`?

trustin commented on core/src/main/java/com/linecorp/armeria/common/http/PathParamExtractor.java:
```diff
@@ -26,38 +28,41 @@
-    static DynamicPath of(String path) {
-        if (!VALID_PATTERN.matcher(path).matches()) {
+    public static PathParamExtractor of(String path) {
+        requireNonNull(path, "path");
```
path -> pathPattern?

trustin commented on core/src/main/java/com/linecorp/armeria/common/http/PathParamExtractor.java:
```diff
@@ -26,38 +28,41 @@
+        if (!VALID_PATTERN.matcher(matchPath).matches()) {
             throw new IllegalArgumentException(
-                    String.format("%s: Invalid DynamicPath (valid: %s)", path, VALID_PATTERN));
+                    String.format("%s: Invalid PathParamExtractor (valid: %s)", path, VALID_PATTERN));
```
`throw new IllegalArgumentException("pathPattern: " + pathPattern);` should be enough.

lespinsideg commented:
@trustin I updated the code. Please review them.

trustin commented on core/src/main/java/com/linecorp/armeria/common/http/PathParamExtractor.java:
```diff
@@ -26,38 +28,24 @@
- *   <li> A set of variables declared in the path.
+ * <li> A regex-compiled form of the path. It is used for matching and extracting.</li>
+ * <li> A skeleton of the path. It is used for duplication detecting.</li>
+ * <li> A set of variables declared in the path.</li>
```
I actually meant:

    <ul>
      <li>A regex ...</li>
      ...
    </ul>

trustin commented:
Thanks a lot, @lespinsideg !
